src/server/transaction.cc

```diff
@@ -1033,31 +1049,45 @@ bool Transaction::ScheduleUniqueShard(EngineShard* shard) {
   auto& sd = shard_data_[SidToId(unique_shard_id_)];
   DCHECK_EQ(TxQueue::kEnd, sd.pq_pos);
 
-  // Fast path - for uncontended keys, just run the callback.
-  // That applies for single key operations like set, get, lpush etc.
-  if (shard->db_slice().CheckLock(mode, lock_args) && shard->shard_lock()->Check(mode)) {
-    RunQuickie(shard);
-    return true;
-  }
+  bool unlocked_keys =
+      shard->db_slice().CheckLock(mode, lock_args) && shard->shard_lock()->Check(mode);
+  bool quick_run = unlocked_keys;
 
-  // we can do it because only a single thread writes into txid_ and sd.
-  txid_ = op_seq.fetch_add(1, memory_order_relaxed);
-  sd.pq_pos = shard->txq()->Insert(this);
+  // Fast path. If none of the keys are locked, we can run briefly atomically on the thread
+  // without acquiring them at all.
+  if (quick_run) {
+    auto result = RunQuickie(shard);
+    local_result_ = result.status;
 
-  DCHECK_EQ(0, sd.local_mask & KEYLOCK_ACQUIRED);
+    if (result.flags & RunnableResult::AVOID_CONCLUDING) {
+      // If we want to run again, we have to actually acquire keys, but keep ourselves disarmed
+      DCHECK_EQ(sd.is_armed, false);
+      unlocked_keys = false;
+    } else {
+      LogAutoJournalOnShard(shard);
+    }
+  }
 
-  shard->db_slice().Acquire(mode, lock_args);
-  sd.local_mask |= KEYLOCK_ACQUIRED;
+  // Slow path. Some of the keys are locked, so we schedule on the transaction queue.
+  if (!unlocked_keys) {
+    coordinator_state_ |= COORD_SCHED;                  // safe because single shard
+    txid_ = op_seq.fetch_add(1, memory_order_relaxed);  // -
+    sd.pq_pos = shard->txq()->Insert(this);
 
-  DVLOG(1) << "Rescheduling " << DebugId() << " into TxQueue of size " << shard->txq()->size();
+    DCHECK_EQ(sd.local_mask & KEYLOCK_ACQUIRED, 0);
+    shard->db_slice().Acquire(mode, lock_args);
+    sd.local_mask |= KEYLOCK_ACQUIRED;
 
-  // If there are blocked transactons waiting for this tx keys, we will add this transaction
-  // to the tx-queue (these keys will be contended). This will happen even if the queue was empty.
-  // In that case we must poll the queue, because there will be no other callback trigerring the
-  // queue before us.
-  shard->PollExecution("schedule_unique", nullptr);
+    DVLOG(1) << "Rescheduling " << DebugId() << " into TxQueue of size " << shard->txq()->size();
 
-  return false;
+    // If there are blocked transactons waiting for this tx keys, we will add this transaction
+    // to the tx-queue (these keys will be contended). This will happen even if the queue was empty.
+    // In that case we must poll the queue, because there will be no other callback trigerring the
+    // queue before us.
+    shard->PollExecution("schedule_unique", nullptr);
+  }
+
+  return quick_run;
```

Comment:
what if quick_run = true but esult.flags is RunnableResult::AVOID_CONCLUDING
should we return false in this case?

Reply:
Yes, because the callback did run and won't run again until requested. This is why quick_run and lock_keys are now different variables. Also I updated the comment above, false means the callback will be run via tx queue